libphobos/libdruntime/gcc/deh.d

```diff
@@ -26,776 +26,927 @@ import gcc.config;
+            eh = cast(ExceptionHeader*) __builtin_calloc(ExceptionHeader.sizeof, 1);
+            // Out of memory while throwing - not much else can be done.
+            if (!eh)
+                terminate(__LINE__);
+        }
+        eh.object = o;
+
+        eh.unwindHeader.exception_class = gdcExceptionClass;
+
+        return eh;
+    }
+
+    /**
+     * Free ExceptionHeader that was created by create().
+     */
+    static void free(ExceptionHeader* eh) @nogc
+    {
+        *eh = ExceptionHeader.init;
+        if (eh != &ehstorage)
+            __builtin_free(eh);
+    }
+
+    /**
+     * Push this onto stack of chained exceptions.
+     */
+    void push() @nogc
+    {
+        next = stack;
+        stack = &this;
+    }
+
+    /**
+     * Pop and return top of chained exception stack.
+     */
+    static ExceptionHeader* pop() @nogc
+    {
+        auto eh = stack;
+        stack = eh.next;
+        return eh;
+    }
+
+    /**
+     * Save stage1 handler information in the exception object.
+     */
+    static void save(_Unwind_Exception* unwindHeader,
+                     _Unwind_Context* context, int handler,
+                     const(ubyte)* lsda, _Unwind_Ptr landingPad)
+    {
+        static if (GNU_ARM_EABI_Unwinder)
+        {
+            unwindHeader.barrier_cache.sp = _Unwind_GetGR(context, UNWIND_STACK_REG);
+            unwindHeader.barrier_cache.bitpattern[1] = cast(_uw)handler;
+            unwindHeader.barrier_cache.bitpattern[2] = cast(_uw)lsda;
+            unwindHeader.barrier_cache.bitpattern[3] = cast(_uw)landingPad;
+        }
+        else
+        {
+            ExceptionHeader* eh = toExceptionHeader(unwindHeader);
+            eh.handler = handler;
+            eh.languageSpecificData = lsda;
+            eh.landingPad = landingPad;
+        }
+    }
+
+    /**
+     * Restore the catch handler data saved during phase1.
+     */
+    static void restore(_Unwind_Exception* unwindHeader, out int handler,
+                        out const(ubyte)* lsda, out _Unwind_Ptr landingPad)
+    {
+        static if (GNU_ARM_EABI_Unwinder)
+        {
+            handler = cast(int)unwindHeader.barrier_cache.bitpattern[1];
+            lsda = cast(ubyte*)unwindHeader.barrier_cache.bitpattern[2];
+            landingPad = cast(_Unwind_Ptr)unwindHeader.barrier_cache.bitpattern[3];
+        }
+        else
+        {
+            ExceptionHeader* eh = toExceptionHeader(unwindHeader);
+            handler = eh.handler;
+            lsda = eh.languageSpecificData;
+            landingPad = cast(_Unwind_Ptr)eh.landingPad;
+        }
+    }
+
+    /**
+     * Look at the chain of inflight exceptions and pick the class type that'll
+     * be looked for in catch clauses.
+     */
+    static getClassInfo(_Unwind_Exception* unwindHeader) @nogc
+    {
+        ExceptionHeader* eh = toExceptionHeader(unwindHeader);
+        Throwable ehobject = eh.object;
+        for (ExceptionHeader* ehn = eh.next; ehn; ehn = ehn.next)
```

Comment:
Right.  Although we do a similar check before installing the handler, this just cuts the middle-man.  For instance in:
```
try
{
    throw new Exception("m2");
}
finally
{
    throw new AnotherException("n2");
}
```
During unwind of the second thrown exception, you'll have get a trace like:
```
[actionTableLookup]: thrown=eh.AnotherException, inspecting=eh.AnotherException
[actionTableLookup]: Saw handler=2
[__gdc_personality]: Replacing handler=2 with 1
[__gdc_personality]: Installing handler=1
```
Whereas this change says, although `AnotherException` was through, we are really looking for the handler for the original `Exception`.
```
[actionTableLookup]: thrown=object.Exception, inspecting=eh.AnotherException
[actionTableLookup]: thrown=object.Exception, inspecting=object.Exception
[actionTableLookup]: Saw handler=1
[__gdc_personality]: Installing handler=1
```